### ISSUE TITLE
Interceptors and unhandled promise rejection

### ISSUE THREAD

**Issue by svpace** [axios/axios#1939]
#### Summary

I am using a interceptor to handle errors globally but I am still getting "Unhandled promise rejection" errors on the console for every request.

``` javascript
axios.interceptors.response.use(
  response => response,
  error => {
    console.log(error.message);
    return Promise.reject(error);
}
----
axios.get('some-url').then(response => /* handle response */);
----
> Error Message
> Unhandled promise rejection Error: "Error message"
```

#### Context

- axios version: 0.18.0
- Environment: Firefox 64, Linux

**Comment by iamtmrobinson** [comment]
You are returning a rejected Promise from your interceptor, but you are not catching it in your calling code.

This should stop the unhandled rejection:
```
axios.get('some-url')
  .then(response => /* handle successful response */)
  .catch(response => /* handle promise rejection */);
```

**Comment by svpace** [comment]
This is exactly what I was trying to avoid. All error handling should be taken care of in the interceptor. I would like to avoid the boilerplate of adding an empty catch in every request.

**Comment by guilhermehn** [comment]
To prevent the catch statement, you should not return `Promise.reject(error)` from the error handler. Replace it with `Promise.resolve()` and it should work as expected.

**Comment by mcstepp** [comment]
I'm getting this same issue, and I *do* have a catch block.

**Comment by agronick** [comment]
@guilhermehn 
Then every request has a result that doesn't match the result it was expecting. I think a lot of people, myself included would like to do something like log the user out on a 401 response and not have it resolve or reject.

**Comment by jbmikk** [comment]
I think the problem is interceptors do their work first and then pass control to the user, and what many people are trying to do is the opposite, trying to catch errors not handled in their user code.

The problem with this is that the only way axios could detect errors not handled by the users is by detecting unhandled promise rejections, which is not very easy.

There is a new browser feature, not yet available in all browsers that would enable to do this, the `unhandledrejection` event, which is already available in Chrome and it is going to be in the next Firefox release, but not yet available in Edge. See more [here](https://caniuse.com/#search=unhandledrejection)

There are two ways this could be done, either completeley implemented by the user, or adding a new type of interceptor to axios, say "user interceptor" or whatever. So when axios detects a a global unhandled rejection it checks to see it is one of its own promises, and then triggers the axios user interceptor. 

I'm not sure it is a good idea to have axios do this or just let users implement it on their own, or using whatever framework they are using. But I think it is important to have a way of managing unhandled errors, because it is the best way of defining global error management. A sort of big try-catch around your whole application, but for promises.

If axios implemented such a feature it would be optional of course, users could still implement their own versions or use another library for global error handling. Angular 2 already has global error management and it works pretty well.

**Comment by chinesedfan** [comment]
> There are two ways this could be done, either completeley implemented by the user, or adding a new type of interceptor to axios, say "user interceptor" or whatever.

Yes, but we want to make axios as simple as possible, unless lots of users request for that option. Feel free to open an issue if you guys persist.

**Comment by brandoncc** [comment]
For anyone else that comes looking, I ended up adding this interceptor:

```
  instance.interceptors.response.use(
    response => response,
    error => {
      // We really want to throw the error so it is handled and we don't get
      // an unhandledrejection error. By throwing here, we are handling the
      // rejection, and bubbling up to the closest error handler (try/catch or
      // catch method call on a promise).
      throw error;
    }
  );
```

**Comment by StallionV** [comment]
Any updates on this? We have exactly the same issue. Once the promise is rejected in the response error do do not want to be adding catch to each request fired. It nullifies the purpose of having a interceptor

**Comment by jbmikk** [comment]
**TL;DR**: Don't use interceptors for default error handling.

I have solved this without using interceptors at all. They are just not useful for handling global errors.
The best way I found to handle global request errors, or any global error for that matter, is simply using the global error handlers for the environment I'm in.
So on error every function in your system should either throw an exception or return a rejected promise.
Depending on your environment you catch these exceptions and rejections globally:

- If you are using vanilla JS: use `window.onerror` for exceptions and `window.onunhandledrejection` for rejections.
- If you are using vue: use `Vue.config.errorHandler` which should be fired if any of your methods, computed values, event handlers either throw an exception or return a rejected promise.
- If you are using any other framework figure out how they handle global errors, if they don't then either the vanilla approach should work or they probably should fix the framework.

This way if your code catches your errors the global error handlers are ignored, but if they don't then they get called. Basically this is the way exceptions were meant to be used in the first place in all languages.

I believe this is the right approach, since it leverages the languages capabilities and is the most standard way of handling errors. If any of the software in the middle does something else with your errors (such as the `beforeRouteEnter` method in the VueRouter which requires doing `next(error)` then you should see how to handle those particular cases).
